File: backend/app/schemas/maintenance.py
from datetime import date as DateType, datetime
from pydantic import BaseModel, ConfigDict


class MaintenanceBase(BaseModel):
    type: str
    date: DateType
    mileage: float | None = None
    cost: float | None = None
    shop_name: str | None = None
    notes: str | None = None
    receipt_path: str | None = None


class MaintenanceCreate(MaintenanceBase):
    vehicle_id: int


class MaintenanceCreateBody(MaintenanceBase):
    """Body for POST; vehicle_id comes from URL."""
    pass


class MaintenanceUpdate(BaseModel):
    type: str | None = None
    date: DateType | None = None
    mileage: float | None = None
    cost: float | None = None
    shop_name: str | None = None
    notes: str | None = None
    receipt_path: str | None = None


class MaintenanceOut(MaintenanceBase):
    id: int
    vehicle_id: int
    created_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/maintenance.ts
import { api, apiFormData } from "./client";

export interface Maintenance {
  id: number;
  vehicle_id: number;
  type: string;
  date: string;
  mileage: number | null;
  cost: number | null;
  shop_name: string | null;
  notes: string | null;
  receipt_path: string | null;
  created_at: string | null;
}

export interface MaintenanceCreate {
  type: string;
  date: string;
  mileage?: number | null;
  cost?: number | null;
  shop_name?: string | null;
  notes?: string | null;
}

export interface MaintenanceUpdate {
  type?: string;
  date?: string;
  mileage?: number | null;
  cost?: number | null;
  shop_name?: string | null;
  notes?: string | null;
}

export async function listMaintenance(vehicleId: number): Promise<Maintenance[]> {
  return api<Maintenance[]>(`/vehicles/${vehicleId}/maintenance`);
}

export async function createMaintenance(vehicleId: number, data: MaintenanceCreate): Promise<Maintenance> {
  return api<Maintenance>(`/vehicles/${vehicleId}/maintenance`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateMaintenance(
  vehicleId: number,
  id: number,
  data: MaintenanceUpdate
): Promise<Maintenance> {
  return api<Maintenance>(`/vehicles/${vehicleId}/maintenance/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

export async function deleteMaintenance(vehicleId: number, id: number): Promise<void> {
  return api<void>(`/vehicles/${vehicleId}/maintenance/${id}`, { method: "DELETE" });
}

export async function uploadMaintenanceReceipt(
  vehicleId: number,
  maintenanceId: number,
  file: File
): Promise<Maintenance> {
  const form = new FormData();
  form.append("file", file);
  return apiFormData<Maintenance>(
    `/vehicles/${vehicleId}/maintenance/${maintenanceId}/receipt`,
    form
  );
}
